NetSDK/Services/SegmentFetcher/Classes/SelfRefreshingSegment.cs

```diff
@@ -1,83 +1,59 @@
 ﻿using log4net;
 using Splitio.Domain;
+using Splitio.Services.Cache.Interfaces;
 using Splitio.Services.Client.Classes;
 using Splitio.Services.SegmentFetcher.Interfaces;
 using System;
+using System.Collections.Concurrent;
 using System.Collections.Generic;
 using System.Linq;
 using System.Text;
 using System.Threading;
+using System.Threading.Tasks;
 
 namespace Splitio.Services.SegmentFetcher.Classes
 {
-    public class SelfRefreshingSegment: Segment
+    public class SelfRefreshingSegment
     {
         private static readonly ILog Log = LogManager.GetLogger(typeof(SelfRefreshingSegment));
+
+        public string name;
+        private SdkReadinessGates gates;
         private ISegmentChangeFetcher segmentChangeFetcher;
-        private int interval;
-        public bool stopped { get; private set; }
+        private ISegmentCache segmentCache;
 
-        public SelfRefreshingSegment(string name, ISegmentChangeFetcher segmentChangeFetcher, SdkReadinessGates gates, int interval, long change_number = -1) : base(name, change_number)
+        public SelfRefreshingSegment(string name, ISegmentChangeFetcher segmentChangeFetcher, SdkReadinessGates gates,  ISegmentCache segmentCache)
         {
+            this.name = name;
             this.segmentChangeFetcher = segmentChangeFetcher;
-            this.interval = interval;
-            this.stopped = true;
+            this.segmentCache = segmentCache;
             this.gates = gates;
         }
 
-        public void Start()
-        {
-            Thread thread = new Thread(StartRefreshing);
-            thread.Start();
-        }
-
-        public void Stop()
-        {
-            stopped = true;
-        }
-
-        private void StartRefreshing()
-        {
-            if (!stopped)
-            {
-                return;
-            }
-
-            stopped = false;
-
-            while (!stopped)
-            {
-                RefreshSegment();
-                Thread.Sleep(interval * 1000);
-            }
-        }
-
-
-        private void RefreshSegment()
+        public void RefreshSegment()
         {        
             while (true)
             {
+                var changeNumber = segmentCache.GetChangeNumber(name);
+
                 try
-                {            
-                    var response = segmentChangeFetcher.Fetch(name, change_number);
+                {
+                    var response = segmentChangeFetcher.Fetch(name, changeNumber);
                     if (response == null)
                     {
-                        return;
+                        break;
                     }
-                    if (change_number >= response.till)
+                    if (changeNumber >= response.till)
```

Comment:
@jwaimann the current segmentChanges service return slices of 2500 elements as a max. If we have 2501, the greedy fetch will do a second request to retrieve the last element, help me to double check we are going to store that final element because I don't remember exactly if the changeNumber will be `changeNumber >= response.till`.